File: docs/public/2025-12-21-leetcoding/dynamic-programming-explained/templates/js/dp-11-edit-distance.js
/*
Title: Edit Distance
Problem: Given two strings, return the minimum number of insertions, deletions, or replacements to convert the first into the second.
Constraints: 0 <= length <= 12, lowercase letters
Examples:
- Input: "horse", "ros" => Output: 3
- Input: "intention", "execution" => Output: 5
- Input: "", "a" => Output: 1
Edge cases: Empty strings.

Test design coverage:
- Boundary tests: smallest sizes, zeros/empty when allowed, single-row/column when applicable.
- Equivalence classes: reachable vs unreachable states, varied distributions of values.
- Pitfall regression: known wrong iteration orders or invalid transitions.
- Randomized small tests with oracle.
*/

function solve(word1, word2) {
  // TODO: implement
  return 0;
}


function refSolve(a, b) {
  const m = a.length;
  const n = b.length;
  const dp = Array.from({ length: m + 1 }, () => Array(n + 1).fill(0));
  for (let i = 0; i <= m; i += 1) dp[i][0] = i;
  for (let j = 0; j <= n; j += 1) dp[0][j] = j;
  for (let i = 1; i <= m; i += 1) {
    for (let j = 1; j <= n; j += 1) {
      if (a[i - 1] === b[j - 1]) dp[i][j] = dp[i - 1][j - 1];
      else dp[i][j] = 1 + Math.min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]);
    }
  }
  return dp[m][n];
}


function assertEqual(actual, expected, message) {
  if (actual !== expected) {
    throw new Error(`Fail: ${message}. Expected ${expected}, got ${actual}`);
  }
}

function assertDeepEqual(actual, expected, message) {
  const a = JSON.stringify(actual);
  const b = JSON.stringify(expected);
  if (a !== b) {
    throw new Error(`Fail: ${message}. Expected ${b}, got ${a}`);
  }
}

let seed = 123456789;
function rand() {
  seed = (seed * 1664525 + 1013904223) >>> 0;
  return seed;
}

function randInt(min, max) {
  return min + (rand() % (max - min + 1));
}

let passed = 0;
let failed = 0;

function run(name, fn) {
  try {
    fn();
    passed += 1;
  } catch (err) {
    failed += 1;
    console.error(`Test failed: ${name}`);
    console.error(err.message);
  }
}

run('deterministic', () => {

  assertEqual(solve("horse", "ros"), 3, 'deterministic');
  assertEqual(solve("intention", "execution"), 5, 'deterministic');
  assertEqual(solve("", "a"), 1, 'deterministic');
  assertEqual(solve("ab", "ab"), 0, 'deterministic');
  assertEqual(solve("kitten", "sitting"), 3, 'deterministic');
  assertEqual(solve("a", ""), 1, 'deterministic');

});

run('random', () => {

  const letters = ['a', 'b', 'c'];
for (let t = 0; t < 20; t += 1) {
  const len1 = randInt(0, 6);
  const len2 = randInt(0, 6);
  const a = Array.from({ length: len1 }, () => letters[randInt(0, 2)]).join('');
  const b = Array.from({ length: len2 }, () => letters[randInt(0, 2)]).join('');
  const expected = refSolve(a, b);
  assertEqual(solve(a, b), expected, `a=${a}, b=${b}`);
}

});

console.log(`Passed: ${passed}, Failed: ${failed}`);

if (failed > 0) process.exit(1);

module.exports = { solve };

File: docs/public/2025-12-21-leetcoding/dynamic-programming-explained/templates/c99/dp-11-edit-distance.c
/*
Title: Edit Distance
Problem: Given two strings, return the minimum number of insertions, deletions, or replacements to convert the first into the second.
Constraints: 0 <= length <= 12, lowercase letters
Examples:
- Input: "horse", "ros" => Output: 3
- Input: "intention", "execution" => Output: 5
- Input: "", "a" => Output: 1
Edge cases: Empty strings.

Test design coverage:
- Boundary tests: smallest sizes, zeros/empty when allowed, single-row/column when applicable.
- Equivalence classes: reachable vs unreachable states, varied distributions of values.
- Pitfall regression: known wrong iteration orders or invalid transitions.
- Randomized small tests with oracle.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int passed = 0;
static int failed = 0;

#define ASSERT_INT_EQ(actual, expected, msg) do { \
  if ((actual) != (expected)) { \
    failed += 1; \
    printf("FAIL: %s (expected %d, got %d)\\n", msg, (expected), (actual)); \
  } else { \
    passed += 1; \
  } \
} while (0)

#define ASSERT_BOOL_EQ(actual, expected, msg) ASSERT_INT_EQ((actual) != 0, (expected) != 0, msg)

static unsigned int seed = 123456789u;
static unsigned int lcg(void) {
  seed = seed * 1664525u + 1013904223u;
  return seed;
}

static int rand_int(int min, int max) {
  return min + (int)(lcg() % (unsigned int)(max - min + 1));
}


int solve(const char *a, const char *b) {
  // TODO: implement
  return 0;
}


static int ref_solve(const char *a, const char *b) {
  int m = (int)strlen(a);
  int n = (int)strlen(b);
  int *dp = (int *)malloc(sizeof(int) * (m + 1) * (n + 1));
  for (int i = 0; i <= m; i += 1) dp[i * (n + 1)] = i;
  for (int j = 0; j <= n; j += 1) dp[j] = j;
  for (int i = 1; i <= m; i += 1) {
    for (int j = 1; j <= n; j += 1) {
      int idx = i * (n + 1) + j;
      if (a[i - 1] == b[j - 1]) dp[idx] = dp[(i - 1) * (n + 1) + j - 1];
      else {
        int del = dp[(i - 1) * (n + 1) + j];
        int ins = dp[i * (n + 1) + j - 1];
        int rep = dp[(i - 1) * (n + 1) + j - 1];
        int best = del < ins ? del : ins;
        best = best < rep ? best : rep;
        dp[idx] = 1 + best;
      }
    }
  }
  int ans = dp[m * (n + 1) + n];
  free(dp);
  return ans;
}


static void test_deterministic(void) {

  ASSERT_INT_EQ(solve("horse", "ros"), 3, "deterministic");
  ASSERT_INT_EQ(solve("intention", "execution"), 5, "deterministic");
  ASSERT_INT_EQ(solve("", "a"), 1, "deterministic");
  ASSERT_INT_EQ(solve("ab", "ab"), 0, "deterministic");
  ASSERT_INT_EQ(solve("kitten", "sitting"), 3, "deterministic");
  ASSERT_INT_EQ(solve("a", ""), 1, "deterministic");

}

static void test_random(void) {

  const char letters[] = "abc";
for (int t = 0; t < 20; t += 1) {
  int len1 = rand_int(0, 6);
  int len2 = rand_int(0, 6);
  char a[8];
  char b[8];
  for (int i = 0; i < len1; i += 1) a[i] = letters[rand_int(0, 2)];
  a[len1] = '\0';
  for (int j = 0; j < len2; j += 1) b[j] = letters[rand_int(0, 2)];
  b[len2] = '\0';
  int expected = ref_solve(a, b);
  ASSERT_INT_EQ(solve(a, b), expected, "random edit");
}

}

int main(void) {

  test_deterministic();

  test_random();

  printf("Passed: %d, Failed: %d\n", passed, failed);

  return failed > 0 ? 1 : 0;

}

